pl_bolts/datamodules/experience_source.py

```diff
@@ -0,0 +1,201 @@
+"""
+Datamodules for RL models that rely on experiences generated during training
+
+Based on implementations found here: https://github.com/Shmuma/ptan/blob/master/ptan/experience.py
+"""
+from collections import deque
+from typing import Iterable, Callable, Tuple, List
+import numpy as np
+import torch
+from gym import Env
+from torch.utils.data import IterableDataset
+
+# Datasets
+from pl_bolts.models.rl.common.agents import Agent
+from pl_bolts.models.rl.common.memory import Experience
+
+
+class ExperienceSourceDataset(IterableDataset):
+    """
+    Basic experience source dataset. Takes a generate_batch function that returns an iterator.
+    The logic for the experience source and how the batch is generated is defined the Lightning model itself
+    """
+
+    def __init__(self, generate_batch: Callable):
+        self.generate_batch = generate_batch
+
+    def __iter__(self) -> Iterable:
+        iterator = self.generate_batch()
+        return iterator
+
+# Experience Sources
+
+
+class ExperienceSource:
+    """
+    Basic single step experience source
+
+    Args:
+        env: Environment that is being used
+        agent: Agent being used to make decisions
+    """
+
+    def __init__(self, env: Env, agent: Agent):
+        self.env = env
+        self.agent = agent
+        self.state = self.env.reset()
+
+    def _reset(self) -> None:
+        """resets the env and state"""
+        self.state = self.env.reset()
+
+    def step(self, device: torch.device) -> Tuple[Experience, float, bool]:
+        """Takes a single step through the environment"""
+        action = self.agent(self.state, device)
+        new_state, reward, done, _ = self.env.step(action)
+        experience = Experience(
+            state=self.state,
+            action=action,
+            reward=reward,
+            new_state=new_state,
+            done=done,
+        )
+        self.state = new_state
+
+        if done:
+            self.state = self.env.reset()
+
+        return experience, reward, done
+
+    def run_episode(self, device: torch.device) -> float:
```

Comment:
is `episode` a common RL term for this? Intuitively I would have called this sequence...

Reply:
It depends on the task. Many tasks are Episodic and have a specified completion time or event. This function was originally used for carrying out a validation episode, but probably isn't needed here